In [39]:
!which python

/home/hienpth/miniconda3/envs/p4ds-env/bin/python


# Final project: Phân tích dữ liệu Data Engineer Jobs
Các thành viên:

1. Phạm Thị Hoài Hiền - 18120178

2. Vương Thị Ngọc Linh - 18120195

In [42]:
import numpy as np 
import matplotlib as plt
import pandas as pd
import re

## Giới thiệu về dữ liệu

Đây là bộ dữ liệu về việc làm Data Engineer và một số vị trí liên quan, link Kaggle lấy dữ liệu: https://www.kaggle.com/andrewmvd/data-engineer-jobs

Datatset này được tạo bởi Picklesueat và nó chứa hơn 2500 công việc liên quan vị trí Data Engineer.

Dataset gồm có 15 cột:
- Job Tile: Tên vị trí
- Salary Estimate: Mức lương
- Job Description: Miêu tả vị trí
- Rating: Xếp hạng công ty tuyển dụng
- Company name: Tên công ty tuyển dụng
- Location: Địa chỉ nơi làm việc
- Headquarters: Trụ sở của công ty
- Size: Quy mô nhân lực của công ty
- Founded: Năm thành lập công ty
- Type of ownership: Quyền sở hữu
- Industry: Lĩnh vực hoạt động của công ty
- Sector: Lĩnh vực chuyên ngành mà công ty phát triển.
- Revenue: Doanh thu mỗi năm của công ty.
- Competitors: Đối thủ cạnh tranh của công ty.
- Easy Apply: Vị trí này có dễ apply không.


## Tiền xử lí

### Đọc dữ liệu vào dataframe

In [43]:
data_engineer_jobs = pd.read_csv("DataEngineer.csv")
data_engineer_jobs.shape

(2528, 15)

Dataset có 2528 dòng và 15 cột, đúng như miêu tả.

In [44]:
data_engineer_jobs.head()

Job Title  \
0                                      Data Engineer   
1  Senior Data Engineer (Healthcare Domain experi...   
2                                     Data Engineers   
3                      Client Trade Support Engineer   
4                                      Data Engineer   

               Salary Estimate  \
0  $80K-$150K (Glassdoor est.)   
1  $80K-$150K (Glassdoor est.)   
2  $80K-$150K (Glassdoor est.)   
3  $80K-$150K (Glassdoor est.)   
4  $80K-$150K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Company Description\nSagence is a management a...     4.5   
1  Key Responsibilities\n\n- Architect, build, an...     3.4   
2  Overview\n\nJob description\n\nPosition Overvi...     5.0   
3  About the Position\n\n\nThis position will sup...     4.8   
4  Data Engineer\n\nJob Details\nLevel\nExperienc...     3.7   

                   Company Name      Location      Headquarters  \
0                  Sagence\n4.5  New York, NY       Chicago, IL   
1   Enterprise Integration\n3.4  New York, NY  Jacksonville, FL   
2     Maestro Technologies\n5.0  New York, NY       Trenton, NJ   
3              Jane Street\n4.8  New York, NY      New York, NY   
4  GNY Insurance Companies\n3.7  New York, NY      New York, NY   

                    Size  Founded  Type of ownership  \
0      1 to 50 employees     2009  Company - Private   
1    51 to 200 employees     1998  Company - Private   
2    51 to 200 employees     2003  Company - Private   
3  501 to 1000 employees     2000  Company - Private   
4   201 to 500 employees     1914  Company - Private   

                                Industry                  Sector  \
0                             Consulting       Business Services   
1                            IT Services  Information Technology   
2                            IT Services  Information Technology   
3  Investment Banking & Asset Management                 Finance   
4                     Insurance Carriers               Insurance   

                      Revenue                       Competitors Easy Apply  
0    $10 to $25 million (USD)               WCI Consulting, PwC         -1  
1    $25 to $50 million (USD)                                -1         -1  
2     $5 to $10 million (USD)                                -1         -1  
3    Unknown / Non-Applicable                                -1         -1  
4  $100 to $500 million (USD)  Travelers, Chubb, Crum & Forster       True

### Xử lí các cột numeric

In [45]:
data_engineer_jobs.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Easy Apply'],
      dtype='object')

Có đầy đủ các cột và nó có tên đúng như mô tả. Tiếp theo ta sẽ xem kiểu dữ liệu của cột.

In [46]:
data_engineer_jobs.dtypes

Job Title             object
Salary Estimate       object
Job Description       object
Rating               float64
Company Name          object
Location              object
Headquarters          object
Size                  object
Founded                int64
Type of ownership     object
Industry              object
Sector                object
Revenue               object
Competitors           object
Easy Apply            object
dtype: object

Theo lí thuyết ta sẽ phải có những cột này là kiểu numeric: `Salary Estimate`, `Rating`, `Founded`, `Size`, `Revenue`. Tuy nhiên theo kết quả trên thì các cột `Salary Estimate`, `Size` và `Revenue` chưa có đúng kiểu dữ liệu (object thay vì int hoặc float). Ta tiến hành đổi kiểu dữ liệu lại cho 3 cột này.

**Đổi kiểu dữ liệu cột `Salary Estimate`**

In [47]:
data_engineer_jobs["Salary Estimate"].unique()

array(['$80K-$150K (Glassdoor est.)', '$73K-$94K (Glassdoor est.)',
       '$85K-$154K (Glassdoor est.)', '$84K-$113K (Glassdoor est.)',
       '$57K-$75K (Glassdoor est.)', '$105K-$121K (Glassdoor est.)',
       '$92K-$117K (Glassdoor est.)', '$88K-$136K (Glassdoor est.)',
       '$95K-$167K (Glassdoor est.)', '$79K-$146K (Glassdoor est.)',
       '$93K-$106K (Glassdoor est.)', '$115K-$201K (Glassdoor est.)',
       '$94K-$164K (Glassdoor est.)', '$49K-$98K (Glassdoor est.)',
       '$56K-$108K (Glassdoor est.)', '$85K-$158K (Glassdoor est.)',
       '$133K-$210K (Glassdoor est.)', '$95K-$99K (Glassdoor est.)',
       '$67K-$72K (Glassdoor est.)', '$53K-$100K (Glassdoor est.)',
       '$81K-$100K (Glassdoor est.)', '$94K-$107K (Glassdoor est.)',
       '$43K-$84K (Glassdoor est.)', '$62K-$92K (Glassdoor est.)',
       '$55K-$106K (Glassdoor est.)', '$69K-$103K (Glassdoor est.)',
       '$54K-$113K (Glassdoor est.)', '$74K-$123K (Glassdoor est.)',
       '$69K-$122K (Glassdoor est.)', 

Tất cả các giá trị đều được biểu diễn trong một khoảng cố định. `Glassdoor est.` là mức lương ước tính trên trang việc làm trực tuyến `Glassdoor`. 
Ta đổi giá trị của cột `Salary Estimate` từ mức lương min-max thành mức lương trung bình để dễ dàng so sánh và tính toán.
Rất may là cột này không có giá trị thiếu.

In [48]:
def find_mean_salary(x):
    S = [int(s) for s in re.findall(r'\d+', x)] #liệt kê (list) tất cả số nguyên trong x
    return (S[0] + S[1]) / 2

data_engineer_jobs["Salary Estimate"] = data_engineer_jobs["Salary Estimate"].apply(find_mean_salary)
data_engineer_jobs = data_engineer_jobs.rename(columns={"Salary Estimate": "Salary est. (thousand USD)"})

**Đổi kiểu dữ liệu cột `Size`**

In [49]:
data_engineer_jobs['Size'].unique()

array(['1 to 50 employees', '51 to 200 employees',
       '501 to 1000 employees', '201 to 500 employees',
       '1001 to 5000 employees', '10000+ employees',
       '5001 to 10000 employees', 'Unknown', '-1'], dtype=object)

Ta thấy chỉ có 7 giá trị phân biệt và 2 giá trị missing (unknown, -1). 
Vậy ta quy đổi thành các giá trị từ 1-7 thể hiện cho quy mô tăng dần như sau:
- Mức độ 1: 1 to 50 employees
- Mức độ 2: 51 to 200 employees
- Mức độ 3: 201 to 500 employees
- Mức độ 4: 501 to 1000 employees
- Mức độ 5: 1001 to 5000 employees
- Mức độ 6: 5001 to 10000 employees
- Mức độ 7: 10000+ employees
Các giá trị -1 và unknown ta sẽ để bằng 0.

In [50]:
def transform_size(x):
    if x == '1 to 50 employees':
        return 1
    elif x == '51 to 200 employees':
        return 2
    elif x == '201 to 500 employees':
        return 2
    elif x == '501 to 1000 employees':
        return 4
    elif x == '1001 to 5000 employees':
        return 5
    elif x == '5001 to 10000 employees':
        return 6
    elif x == '10000+ employees':
        return 7
    else:
        return -1

data_engineer_jobs['Size'] = data_engineer_jobs['Size'].apply(transform_size)

**Đổi kiểu dữ liệu cột `Revenue`**

In [51]:
data_engineer_jobs['Revenue'].unique()

array(['$10 to $25 million (USD)', '$25 to $50 million (USD)',
       '$5 to $10 million (USD)', 'Unknown / Non-Applicable',
       '$100 to $500 million (USD)', '$2 to $5 billion (USD)',
       '$500 million to $1 billion (USD)', '$50 to $100 million (USD)',
       '$10+ billion (USD)', '$1 to $2 billion (USD)',
       '$1 to $5 million (USD)', 'Less than $1 million (USD)',
       '$5 to $10 billion (USD)', '-1'], dtype=object)

Ta thấy có rất nhiều cách biểu diễn khác nhau cho cột này (có khoảng from..to, có con số X+, có less than). 
Do số tiền chỉ có đơn vị là triệu (million) và tỷ (billion) nên ta sẽ đổi hết giá trị về thành đơn vị tỉ đô
Đổi giá trị cột `Revenue` từ các khoảng doanh thu bằng `giá trị cao nhất (max)` trong khoảng:
- Các dòng có giá trị X+ sẽ được quy ước là X+1
- Các dòng có giá trị Less than X quy ước là X 
- Missing value sẽ được quy ước là 0

In [52]:
def find_max_revenue(x):
    if x == 'Unknown / Non-Applicable' or x == '-1':
        return -1
    else:
        S = [int(s) for s in re.findall(r'\d+', x)]
        if 'billion' in x and 'million' not in x:
            S = [s*1000 for s in S]
        if '+' in x:
            S.append(S[0] + 1)
        if 'Less than' in x:
            S.append(S[0])
            S[0] = 0
        return S[1]
data_engineer_jobs['Revenue'] = data_engineer_jobs['Revenue'].apply(find_max_revenue)
data_engineer_jobs = data_engineer_jobs.rename(columns={'Revenue': 'Revenue (million USD)'})

In [53]:
data_engineer_jobs

Job Title  \
0                                         Data Engineer   
1     Senior Data Engineer (Healthcare Domain experi...   
2                                        Data Engineers   
3                         Client Trade Support Engineer   
4                                         Data Engineer   
...                                                 ...   
2523                                     Cloud Engineer   
2524                                Electrical Engineer   
2525                              Sr. Big Data Engineer   
2526                     Senior Big Data / ETL Engineer   
2527                                     Cloud Engineer   

      Salary est. (thousand USD)  \
0                          115.0   
1                          115.0   
2                          115.0   
3                          115.0   
4                          115.0   
...                          ...   
2523                       102.0   
2524                       102.0   
2525                       102.0   
2526                       102.0   
2527                       102.0   

                                        Job Description  Rating  \
0     Company Description\nSagence is a management a...     4.5   
1     Key Responsibilities\n\n- Architect, build, an...     3.4   
2     Overview\n\nJob description\n\nPosition Overvi...     5.0   
3     About the Position\n\n\nThis position will sup...     4.8   
4     Data Engineer\n\nJob Details\nLevel\nExperienc...     3.7   
...                                                 ...     ...   
2523  The Expertise You Have Proven expertise in dev...     4.5   
2524  Lockheed Martin is seeking a Circuit and FPGA ...     2.9   
2525  Posting Title Principal Data Engineer (Contrac...     3.7   
2526  Your Opportunity\n\nCharles Schwab & Co., Inc ...     3.5   
2527  Cloud Engineer Westlake, TX 6+ months Descript...     3.6   

                                   Company Name        Location  \
0                                  Sagence\n4.5    New York, NY   
1                   Enterprise Integration\n3.4    New York, NY   
2                     Maestro Technologies\n5.0    New York, NY   
3                              Jane Street\n4.8    New York, NY   
4                  GNY Insurance Companies\n3.7    New York, NY   
...                                         ...             ...   
2523                            PMG Global\n4.5    Westlake, TX   
2524            Albin Engineering Services\n2.9  Fort Worth, TX   
2525  Cincinnati Bell Technology Solutions\n3.7      Irving, TX   
2526                                Schwab\n3.5    Westlake, TX   
2527        SPECTRAFORCE TECHNOLOGIES Inc.\n3.6    Westlake, TX   

             Headquarters  Size  Founded  Type of ownership  \
0             Chicago, IL     1     2009  Company - Private   
1        Jacksonville, FL     2     1998  Company - Private   
2             Trenton, NJ     2     2003  Company - Private   
3            New York, NY     4     2000  Company - Private   
4            New York, NY     2     1914  Company - Private   
...                   ...   ...      ...                ...   
2523          Herndon, VA     1       -1  Company - Private   
2524      Santa Clara, CA     2     1993  Company - Private   
2525       Cincinnati, OH     4       -1  Company - Private   
2526  Dettenheim, Germany    -1       -1   Company - Public   
2527          Raleigh, NC     4     2000  Company - Private   

                                   Industry                  Sector  \
0                                Consulting       Business Services   
1                               IT Services  Information Technology   
2                               IT Services  Information Technology   
3     Investment Banking & Asset Management                 Finance   
4                        Insurance Carriers               Insurance   
...                                     ...                     ...   
2523           Computer Hardware & Softw

### Xử lí các cột category

Ở cột `Company Name`, ta thấy có phần "\n rating" phía sau tên công ty là dư thừa (vì đã có cột `rating` riêng), vì vậy ta xoá nó đi.

In [54]:
data_engineer_jobs['Company Name'] = data_engineer_jobs['Company Name'].str.partition("\n")

### Xử lí missing value

In [55]:
data_engineer_jobs.isnull().sum() / len(data_engineer_jobs) * 100

Job Title                     0.0
Salary est. (thousand USD)    0.0
Job Description               0.0
Rating                        0.0
Company Name                  0.0
Location                      0.0
Headquarters                  0.0
Size                          0.0
Founded                       0.0
Type of ownership             0.0
Industry                      0.0
Sector                        0.0
Revenue (million USD)         0.0
Competitors                   0.0
Easy Apply                    0.0
dtype: float64

Tỉ lệ missing value của tưng cột là 0.0, tuy nhiên sau khi xem dữ liệu, có thể thấy tác giả đã thay các giá trị thiếu thành `-1` và `Unknown / Non-Applicable`.
Phần `Unknown / Non-Applicable` ta đã xử lí ở cột `Revenue` rồi nên bây giờ chỉ xem xét tới giá trị -1 ở các cột.

In [56]:
def find_missing_value(col):
    if col.dtypes == np.dtype('int64') or col.dtypes == np.dtype('float64'):
        return abs(col[col == -1].sum() / len(col) * 100)
    else:
        return len(col[col == '-1']) / len(col) * 100
missing_ratio = data_engineer_jobs.apply(find_missing_value)
missing_ratio

Job Title                      0.000000
Salary est. (thousand USD)     0.000000
Job Description                0.000000
Rating                         9.256329
Company Name                   0.000000
Location                       0.000000
Headquarters                   4.905063
Size                           7.199367
Founded                       22.587025
Type of ownership              4.707278
Industry                      14.952532
Sector                        14.833861
Revenue (million USD)         32.950949
Competitors                   69.185127
Easy Apply                    94.224684
dtype: float64

Cột Competitors và Easy Apply có rất nhiều missing value, và dữ liệu hai cột này cũng không cần thiết cho câu hỏi, vì vậy sẽ xóa hai cột này để làm sạch dữ liệu.

In [57]:
data_engineer_jobs = data_engineer_jobs.drop(columns=['Competitors', 'Easy Apply'])

Đối với các cột có tỉ lệ missing cao (`Industry`, `Sector`, `Revenue`, `Founded`), ta sẽ xử lí nó theo từng câu hỏi đặt ra ở phần sau.

In [ ]:
company = data_engineer_jobs[data_engineer_jobs['Revenue (million USD)'] == -1]
company = company.groupby(['Company Name']).size()
company[company > 1]

In [ ]:
tmp = data_engineer_jobs[data_engineer_jobs['Company Name'] == 'Akuna Capital'] 
tmp[['Location','Revenue (million USD)']]

## Các câu hỏi cần trả lời

(chém gió vô đây)